File: src/main/java/com/pixelreel/service/provider/DoubanMovieSearchProvider.java
package com.pixelreel.service.provider;

import com.pixelreel.config.DoubanProperties;
import com.pixelreel.dto.douban.DoubanSearchResponse;
import com.pixelreel.dto.douban.DoubanSubject;
import com.pixelreel.dto.external.ExternalMovieSearchResult;
import com.pixelreel.dto.external.MovieRecordSuggestion;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.entity.Movie;
import com.pixelreel.enums.RecordStatus;
import com.pixelreel.service.MovieService;
import com.pixelreel.service.search.MovieSearchProvider;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class DoubanMovieSearchProvider implements MovieSearchProvider {
  private static final int DEFAULT_PAGE_SIZE = 20;

  private final MovieService movieService;

  @Qualifier("doubanRestClient")
  private final RestClient doubanRestClient;

  private final DoubanProperties doubanProperties;

  @Override
  public String id() {
    return "douban";
  }

  @Override
  public ProviderSearchResult<ExternalMovieSearchResult> search(String query, int page) {
    ProviderSearchResult<ExternalMovieSearchResult> result = new ProviderSearchResult<>();
    result.setProvider(id());

    if (!doubanProperties.enabled()) {
      result.setEnabled(false);
      result.setMessage("豆瓣未启用");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(query)) {
      throw new IllegalArgumentException("query must not be blank");
    }

    int normalizedPage = Math.max(page, 1);
    int start = (normalizedPage - 1) * DEFAULT_PAGE_SIZE;

    DoubanSearchResponse response = doubanRestClient
        .get()
        .uri(uriBuilder -> uriBuilder
            .path("/movie/search")
            .queryParam("q", query)
            .queryParam("start", start)
            .queryParam("count", DEFAULT_PAGE_SIZE)
            .queryParamIfPresent("apikey", StringUtils.hasText(doubanProperties.apiKey())
                ? java.util.Optional.of(doubanProperties.apiKey())
                : java.util.Optional.empty())
            .build())
        .retrieve()
        .body(DoubanSearchResponse.class);

    List<DoubanSubject> subjects = response == null ? Collections.emptyList() : response.getSubjects();
    Map<String, Movie> existingByDoubanId = findExistingByDoubanId(subjects);

    List<ExternalMovieSearchResult> results = new ArrayList<>();
    for (DoubanSubject subject : subjects) {
      ExternalMovieSearchResult mapped = new ExternalMovieSearchResult();
      mapped.setProvider(id());
      mapped.setDoubanId(subject.getId());
      mapped.setTitle(StringUtils.hasText(subject.getTitle()) ? subject.getTitle() : subject.getOriginalTitle());
      mapped.setReleaseDate(subject.getYear());
      mapped.setOverview(subject.getSummary());
      mapped.setPosterUrl(subject.getImages() == null ? null : subject.getImages().getLarge());

      Movie existing = subject.getId() == null ? null : existingByDoubanId.get(subject.getId());
      mapped.setAlreadyAdded(existing != null);
      mapped.setExistingRecordId(existing == null ? null : existing.getId());
      mapped.setSuggestedRecord(buildSuggestion(mapped));
      results.add(mapped);
    }

    int total = response == null || response.getTotal() == null ? results.size() : response.getTotal();
    int totalPages = total == 0 ? 0 : (int) Math.ceil(total / (double) DEFAULT_PAGE_SIZE);

    result.setEnabled(true);
    result.setPage(normalizedPage);
    result.setTotalPages(totalPages);
    result.setTotalResults(total);
    result.setResults(results);
    return result;
  }

  private Map<String, Movie> findExistingByDoubanId(List<DoubanSubject> subjects) {
    List<String> ids = subjects.stream()
        .map(DoubanSubject::getId)
        .filter(StringUtils::hasText)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return movieService.lambdaQuery()
        .in(Movie::getDoubanId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Movie::getDoubanId, Function.identity(), (a, b) -> a));
  }

  private MovieRecordSuggestion buildSuggestion(ExternalMovieSearchResult mapped) {
    MovieRecordSuggestion suggestion = new MovieRecordSuggestion();
    suggestion.setDoubanId(mapped.getDoubanId());
    suggestion.setTitle(mapped.getTitle());
    suggestion.setPosterUrl(mapped.getPosterUrl());
    suggestion.setStatus(RecordStatus.WANT);
    suggestion.setRating(null);
    suggestion.setShortReview("");
    return suggestion;
  }
}

File: express-backend/src/services/provider/DoubanMovieSearchProvider.ts
import axios from 'axios';
import { config } from '../../config';
import { prisma } from '../../config/db';
import { MovieSearchProvider } from '../provider/MovieSearchProvider';
import {
  ExternalMovieSearchResult,
  ProviderSearchResult,
  MovieRecordSuggestion,
} from '../../dto/external-search';
import { RecordStatus } from '../../enums/RecordStatus';

// 豆瓣影视搜索 Provider，与 Java 端 DoubanMovieSearchProvider 完全对齐
export class DoubanMovieSearchProvider implements MovieSearchProvider {
  private static readonly PAGE_SIZE = 20;

  id(): string {
    return 'douban';
  }

  async search(query: string, page: number): Promise<ProviderSearchResult<ExternalMovieSearchResult>> {
    const result: ProviderSearchResult<ExternalMovieSearchResult> = {
      provider: this.id(),
      enabled: true,
      message: '',
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };

    if (!config.douban.cookie) {
      result.enabled = false;
      result.message = '豆瓣未启用';
      return result;
    }

    if (!query) throw new Error('query must not be blank');
    const normalizedPage = Math.max(page, 1);
    const start = (normalizedPage - 1) * DoubanMovieSearchProvider.PAGE_SIZE;

    const response = await axios.get(`${config.douban.baseUrl}/movie/search`, {
      params: { q: query, start, count: DoubanMovieSearchProvider.PAGE_SIZE },
      headers: { Cookie: config.douban.cookie },
    });

    const subjects = response.data?.subjects ?? [];
    const doubanIds = subjects.map((s: any) => s.id).filter(Boolean);
    const existingMap = doubanIds.length > 0
      ? await this.findExistingByDoubanId(doubanIds)
      : new Map<string, any>();

    const results: ExternalMovieSearchResult[] = subjects.map((subject: any) => {
      const existing = subject.id ? existingMap.get(subject.id) ?? null : null;
      const mapped: ExternalMovieSearchResult = {
        provider: this.id(),
        tmdbId: null,
        imdbId: null,
        doubanId: subject.id ?? null,
        traktId: null,
        title: subject.title || subject.original_title || '',
        posterUrl: subject.images?.large ?? null,
        releaseDate: subject.year ?? null,
        overview: subject.summary ?? null,
        alreadyAdded: existing !== null,
        existingRecordId: existing?.id != null ? Number(existing.id) : null,
        suggestedRecord: null,
      };
      mapped.suggestedRecord = this.buildSuggestion(mapped);
      return mapped;
    });

    const total = response.data?.total ?? results.length;
    const totalPages = total === 0 ? 0 : Math.ceil(total / DoubanMovieSearchProvider.PAGE_SIZE);

    result.page = normalizedPage;
    result.totalPages = totalPages;
    result.totalResults = total;
    result.results = results;
    return result;
  }

  private async findExistingByDoubanId(ids: string[]): Promise<Map<string, any>> {
    const movies = await prisma.movie.findMany({ where: { doubanId: { in: ids } } });
    return new Map(movies.map((m) => [m.doubanId!, m]));
  }

  private buildSuggestion(mapped: ExternalMovieSearchResult): MovieRecordSuggestion {
    return {
      tmdbId: mapped.tmdbId,
      imdbId: mapped.imdbId,
      doubanId: mapped.doubanId,
      traktId: mapped.traktId,
      title: mapped.title,
      posterUrl: mapped.posterUrl,
      status: RecordStatus.WANT,
      rating: null,
      shortReview: '',
    };
  }
}